ddtrace/tracer/spancontext.go

```diff
@@ -246,11 +267,23 @@ func (t *trace) finishedOne(s *span) {
 	if len(t.spans) != t.finished {
 		return
 	}
-	if tr, ok := internal.GetGlobalTracer().(*tracer); ok {
-		// we have a tracer that can receive completed traces.
-		tr.pushTrace(t.spans)
-		atomic.AddInt64(&tr.spansFinished, int64(len(t.spans)))
+	defer func() {
+		t.spans = nil
+		t.finished = 0 // important, because a buffer can be used for several flushes
+	}()
+	tr, ok := internal.GetGlobalTracer().(*tracer)
+	if !ok {
+		return
+	}
+	// we have a tracer that can receive completed traces.
+	atomic.AddInt64(&tr.spansFinished, int64(len(t.spans)))
+	sd := samplingDecision(atomic.LoadInt64((*int64)(&t.samplingDecision)))
+	if sd != decisionKeep {
+		if t.priority != nil && *(t.priority) == ext.PriorityAutoReject {
```

Comment:
Can we please use the dedicated method for this? https://github.com/DataDog/dd-trace-go/pull/963#discussion_r671440098

We want a unified experience. If the priority retrieval logic ever changes, we want a single source of truth. With the current version, that situation can cause a bug.

Reply:
I can't because there is a lock in the method.

Reply:
I could add a new method with no lock, and have the lock method call the method with no lock if you prefer.

Reply:
I will leave it up to you to decide. No strong feelings here.